File: PYTHON/Pertemuan 01/Mahasiswa.py
import sys
sys.stdout.reconfigure(encoding='utf-8')
class Mahasiswa:
    def __init__(self, nama: str, nim: str, ipk: float):
        self._nama = nama
        self._nim = nim
        self._ipk = 0.0
        self.ipk = ipk  # uses setter for validation

    @property
    def nama(self) -> str:
        return self._nama

    @property
    def nim(self) -> str:
        return self._nim

    @property
    def ipk(self) -> float:
        return self._ipk

    @ipk.setter
    def ipk(self, value: float):
        if 0.0 <= value <= 4.0:
            self._ipk = value
        else:
            print(f"⚠  IPK {value} tidak valid! Harus antara 0.0–4.0. IPK tidak diubah.")

    def belajar(self):
        print(f"{self.nama} sedang belajar.")

    def hadir_kuliah(self, mata_kuliah: str):
        print(f"{self.nama} hadir di kuliah: {mata_kuliah}")

    def tampilkan_info(self):
        print(f"  Nama : {self.nama:<20} | NIM : {self.nim:<12} | IPK : {self.ipk:.2f}")

File: JAVA/Pertemuan 01/src/id/ac/uph/pbo/pertemuan01/Mahasiswa.java
package id.ac.uph.pbo.pertemuan01;

public class Mahasiswa {

    // Atribut bersifat private → tidak bisa diakses langsung dari luar class
    private String nama;
    private String nim;
    private double ipk;

    // ── Constructor: membuat objek Mahasiswa baru ────────────────────────────
    public Mahasiswa(String nama, String nim, double ipk) {
        this.nama = nama;
        this.nim  = nim;
        this.setIpk(ipk);   // melalui setter agar tervalidasi
    }

    // ── Setter IPK dengan validasi ───────────────────────────────────────────
    public void setIpk(double ipk) {
        if (ipk >= 0.0 && ipk <= 4.0) {
            this.ipk = ipk;
        } else {
            System.out.println("⚠  IPK " + ipk + " tidak valid! Harus antara 0.0–4.0."
                    + " IPK tidak diubah.");
        }
    }

    // ── Getter ───────────────────────────────────────────────────────────────
    public String getNama()  { return nama; }
    public String getNim()   { return nim;  }
    public double getIpk()   { return ipk;  }

    // ── Method perilaku ──────────────────────────────────────────────────────
    public void belajar() {
        System.out.println(nama + " sedang belajar.");
    }

    public void hadirKuliah(String mataKuliah) {
        System.out.println(nama + " hadir di kuliah: " + mataKuliah);
    }

    // ── Tampilkan info singkat ───────────────────────────────────────────────
    public void tampilkanInfo() {
        System.out.printf("  Nama : %-20s | NIM : %-12s | IPK : %.2f%n",
                nama, nim, ipk);
    }
}
